File: src/duElementAPI/toggleUnit.lua
--[[    
    toggleUnit.lua 
    Copyright (C) 2020 Stephane Boivin (Devgeek studio enr.)
    
    This file is part of "DU offline sandbox API".

    "DU offline sandbox API" is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    "DU offline sandbox API" is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with "DU offline sandbox API".  If not, see <https://www.gnu.org/licenses/>.
]]
Itoggleunit = {id, name, type}

function Itoggleunit:new(o)
  o = o or {}
  setmetatable(o, self)
  self.__index = self

  o.maxHitPoints = 2000
  o.hitPoints = 2000
  o.mass = 28.67325
  o.class = 'ManualSwitchUnit'
  o.integrity = 100

  o.toggleunit = ToggleUnit:new{id = o.id, name = o.name, type = o.type}
  o.generic = genericElement:new{id = o.id, maxHitPoints = o.maxHitPoints, hitPoints = o.hitPoints, mass = o.mass, class = o.class, integrity = o.integrity }

  return o
end

function createInterfaceToggleUnit(id, name, type)

  local obj = Itoggleunit:new{id=id,name=name,type=type}
  obj.activate = function() obj.toggleunit:activate() end
  obj.deactivate = function() obj.toggleunit:deactivate() end
  obj.toggle = function() obj.toggleunit:toggle() end
  obj.getState = function() return obj.toggleunit:getState() end

  obj.hide = function() obj.generic:hide() end
  obj.show = function() obj.generic:show() end
  obj.getData = function() obj.generic:getData() end
  obj.getId = function() return obj.generic:getId() end
  obj.getMass = function() return obj.generic:getMass() end
  obj.getElementClass = function() return obj.generic:getElementClass() end
  obj.getIntegrity = function() return obj.generic:getIntegrity() end
  obj.getHitPoints = function() return obj.generic:getHitPoints() end
  obj.getMaxHitPoints = function() return obj.generic:getMaxHitPoints() end
  obj._setMass = function(m) obj.generic:setMass(m) end
  obj._setMaxHitPoints = function(mhp) obj.generic:setMaxHitPoints(mhp) end
  obj._setHitPoints = function(hp) obj.generic:setHitPoints(hp) end
  obj._setIntegrity = function(int) obj.generic:setIntegrity(int) end
   
  return obj
end

-- State 0 = closed
-- State 1 = open
ToggleUnit = {id, State = 0, name = "ManualSwitchUnit"}

function ToggleUnit:new(o)
  o = o or {}
  setmetatable(o, self)
  self.__index = self
  o.State = 0;
  if _verboseLUA == 1 then print("[LUA] "..o.type.." ["..o.name.."] created with the id ["..o.id.."]") end  
  return o
end

function ToggleUnit:activate()
  self.State = 1;
  if _verboseLUA == 1 then print("[LUA] "..self.type.." ["..self.name.."] opened") end
  JavaWindow:setState(self.id, {self.State,self.name})    
end

function ToggleUnit:deactivate()
  self.State = 0;
  if _verboseLUA == 1 then print("[LUA] "..self.type.." ["..self.name.."] closed") end
  JavaWindow:setState(self.id, {self.State,self.name})    
end

function ToggleUnit:toggle()
  if self.State == 0 then
    self:activate()
  else
    self:deactivate()
  end  
end

function ToggleUnit:getState()
  return self.State
end

File: src/duJavaAPI/ToggleUnit.java
/*    
    ToggleUnit.java 
     Copyright (C) 2021 Stephane Boivin (Discord: Nmare418#6397)
    
    This file is part of "DU lua sandbox API".

    "DU lua sandbox API" is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    "DU lua sandbox API" is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with "DU lua sandbox API".  If not, see <https://www.gnu.org/licenses/>.
*/
package duJavaAPI;

/*
 ****************************************************************************************
 * Used for Force fields, ship doors, landing gears, manual switches and light squares  * 
 ****************************************************************************************

  use setPictures(String disabledPicture, String activePicture ) to change the pictures
		
*/	

import java.awt.Color;
import java.awt.Image;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

import javax.swing.JPanel;
import javax.swing.border.LineBorder;

import org.luaj.vm2.LuaTable;
import org.luaj.vm2.lib.jse.CoerceJavaToLua;

import javax.swing.JLabel;
import javax.swing.ImageIcon;

public class ToggleUnit extends BaseElement {
    private String disabledPicture = "src/pictures/porte0.png";
    private String activePicture = "src/pictures/porte1.png";
 	private JLabel lblPic = new JLabel();
	public int status = 0;
	public String type = "";
    int sizeX = 100;
    int sizeY = 100;
    private int picX = 2;
    private int picY = 2;
    private int picSizeX = 93;
    private int picSizeY = 75;
    
	// types:  Door, ForceField, LandingGear, Switch, Light 
	public ToggleUnit(int pid, String ptype, String pname, int px, int py, boolean pverboseJava) {
		
		name = pname;
		x = px;
		y = py;
		id = pid;
		type = ptype;
		verboseJava = pverboseJava;
		
		switch (ptype) {
	      case "Door":					
		    disabledPicture = "src/pictures/elements/Door_Closed.png";
		    activePicture = "src/pictures/elements/Door_Open.png";
		    break;
		  case "ForceField":					
		    disabledPicture = "src/pictures/elements/Door_Closed.png";
		    activePicture = "src/pictures/elements/Door_Open.png";
		    break;
		  case "LandingGear":					
		    disabledPicture = "src/pictures/elements/LandingGear_Off.png";
		    activePicture = "src/pictures/elements/LandingGear_On.png";
		    break;
		  case "Switch":
		    disabledPicture = "src/pictures/elements/ToggleButton_Off.png";
		    activePicture = "src/pictures/elements/ToggleButton_On.png";
		    break;
		  case "Light":
		    disabledPicture = "src/pictures/elements/Light_Off.png";
		    activePicture = "src/pictures/elements/Light_On.png";
		}
		
		// create panel
		panel = new JPanel();		
		panel.setLayout(null);
		panel.setBorder(LineBorder.createBlackLineBorder());
		panel.setBackground(Color.black);				
		panel.setBounds(x, y, sizeX, sizeY);
		
		// white frame
		JPanel lblPicWhite = new JPanel();		
	 	lblPicWhite.setBounds(1, 1, sizeX-3, sizeY-3);
	 	lblPicWhite.setBackground(Color.white);
		panel.add(lblPicWhite, 1, 0);
	
		// picture
	 	JLabel lblPic = new JLabel();
		if(status == 0) {
	 	    setPicture(lblPic, disabledPicture, picX, picY, picSizeX, picSizeY);
		} else {
	 	    setPicture(lblPic, activePicture, picX, picY, picSizeX, picSizeY);
		}
	    lblPic.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
            	sandbox.RUN(name+".toggle()", name+":Manual activation");
            }

        });		
		panel.add(lblPic, 1, 0);
		
		// lua require and interface 
		urlAPI = "src/duElementAPI/toggleUnit.lua";
        luaCall = pname+" = createInterfaceToggleUnit(\""+pid+"\", \""+pname+"\", \""+ptype+"\")";
		
        if(verboseJava) System.out.println("[JAVA] "+type+" ["+pname+"] created");	 		
	}


	@Override	
    public Object get(String[] param) {    	
		String pcommand = param[0];
		if(pcommand.equals("getState")) { // internal function
			return status; 
	    }		
		if(pcommand.equals("getType")) { // internal function
			return type; 
	    }		
		return -1;
    }
	@Override
	public void update(String[] param) {
		// {State,name}
		panel.remove(lblPic);
		JLabel lblPic = new JLabel();
		if(Integer.valueOf(param[0]) == 0) {
			   status = 0;
			   setPicture(lblPic, disabledPicture, picX, picY, picSizeX, picSizeY);
		} else {
			   status = 1;
			   setPicture(lblPic, activePicture, picX, picY, picSizeX, picSizeY);
		}
	    lblPic.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
            	sandbox.RUN(name+".toggle()", name+":Manual activation");
            }

        });			
		panel.add(lblPic, 1, 0);
		panel.revalidate();
	    panel.repaint(); 
		if(verboseJava) System.out.println("[JAVA] "+type+" " + param[1] + " state set to  " + param[0]); 
	}
}
